Makefile

```diff
@@ -0,0 +1,50 @@
+NAME=footbot
+
+.PHONY: bash
+bash:
+	docker-compose run --service-ports --name $(NAME) $(NAME) bash
```

Comment:
why do we use `docker-compose` here not just `docker`?

Reply:
`docker-compose` has a couple of benefits:
- with the `docker-compose.yaml`, you can specify all the volumes, ports, etc in one place, rather than repeating this in every `docker run` call
- `docker-compose run` automatically builds the image if it doesn't exist, which is handy

Reply:
it builds the image if it doesn't exist already, but if the image already exist and is out-of-date, it still uses that- 

```
 grw@pc  ~/s/footbot  remotes/upstream/docker ⁝ S ✱ ? 
 ❯ echo "some-package" >> requirements.txt                                                                                                                         stdenv

 grw@pc  ~/s/footbot  remotes/upstream/docker ⁝ S ✱ ? 
 ❯ make serve                                                                                                                                                      stdenv
make clean-container
make[1]: Entering directory '/home/grw/src/footbot'
docker rm -f footbot
footbot
make[1]: Leaving directory '/home/grw/src/footbot'
docker-compose run --service-ports --name footbot footbot
```

^ this should have re-built the image

maybe just a case of adding `build` as a dependency of  `serve`/`test`/`shell` etc

Reply:
Sure, I'll change that now

Reply:
each command in the dockerfile is cached as a layer, so if requirements.txt hasn't change since last time you ran it the build should run straight away

Reply:
The problem is that it doesn't cache the following:
`RUN apt-get update && apt-get install -y build-essential`

As a result, none of the subsequent layers are cached, so it reinstalls all the requirements again, alongside the build-essential stuff, which can take a couple of minutes.

Reply:
what? why would that layer not be cached?

```
Sending build context to Docker daemon  110.4MB

Step 1/7 : FROM python:3.8-slim-buster
 ---> 8671a55eb48a
Step 2/7 : WORKDIR /app
 ---> Using cache
 ---> e9c2b5d52aa2
Step 3/7 : RUN apt-get update && apt-get install -y build-essential
 ---> Using cache
 ---> 2cbdde46f2f4
Step 4/7 : COPY requirements.txt requirements.txt
 ---> Using cache
 ---> 22476ae37ecf
Step 5/7 : RUN pip install -r requirements.txt
 ---> Running in 45366ebe6ab8

```

Reply:
Weird... caching for me now. Perhaps I got unlucky earlier with an apt-get update?
In any case, I'll add `clean build` as deps for `bash` and `serve` as suggested

Reply:
changing the packages to install invalidated your cache i imagine